cuda_core/cuda/core/experimental/_linker.py

```diff
@@ -0,0 +1,309 @@
+            self.formatted_options.append(f"-prec-sqrt={'true' if self.prec_sqrt else 'false'}")
+        if self.fma is not None:
+            self.formatted_options.append(f"-fma={'true' if self.fma else 'false'}")
+        if self.kernels_used is not None:
+            for kernel in self.kernels_used:
+                self.formatted_options.append(f"-kernels-used={kernel}")
+        if self.variables_used is not None:
+            for variable in self.variables_used:
+                self.formatted_options.append(f"-variables-used={variable}")
+        if self.optimize_unused_variables is not None:
+            self.formatted_options.append("-optimize-unused-variables")
+        if self.xptxas is not None:
+            for opt in self.xptxas:
+                self.formatted_options.append(f"-Xptxas={opt}")
+        if self.split_compile is not None:
+            self.formatted_options.append(f"-split-compile={self.split_compile}")
+        if self.split_compile_extended is not None:
+            self.formatted_options.append(f"-split-compile-extended={self.split_compile_extended}")
+        if self.no_cache is not None:
+            self.formatted_options.append("-no-cache")
+
+
+class Linker:
+    """
+    Linker class for managing the linking of object codes with specified options.
+
+    Parameters
+    ----------
+    object_codes : ObjectCode
+        One or more ObjectCode objects to be linked.
+    options : LinkerOptions, optional
+        Options for the linker. If not provided, default options will be used.
+
+    Attributes
+    ----------
+    _options : LinkerOptions
+        The options used for the linker.
+    _handle : handle
+        The handle to the linker created by nvjitlink.
+
+    Methods
+    -------
+    _add_code_object(object_code)
+        Adds an object code to the linker.
+    close()
+        Closes the linker and releases resources.
+    """
+
+    class _MembersNeededForFinalize:
+        __slots__ = ("handle",)
+
+        def __init__(self, program_obj, handle):
+            self.handle = handle
+            weakref.finalize(program_obj, self.close)
+
+        def close(self):
+            if self.handle is not None:
+                nvjitlink.destroy(self.handle)
+                self.handle = None
+
+    __slots__ = ("__weakref__", "_mnff", "_options")
+
+    def __init__(self, *object_codes: ObjectCode, options: LinkerOptions = None):
+        self._options = options = check_or_create_options(LinkerOptions, options, "Linker options")
+        self._mnff.handle = nvjitlink.create(len(options.formatted_options), options.formatted_options)
+        self._mnff = Linker._MembersNeededForFinalize(self, None)
+
+        if len(object_codes) == 0:
+            raise ValueError("At least one ObjectCode object must be provided")
+
+        for code in object_codes:
+            assert isinstance(code, ObjectCode)
+            self._add_code_object(code)
+
+        weakref.finalize(self, self.close)

```

Comment:
This needs to be deleted (it creates a reference cycle and will keep `self` alive until the start of the Python interpreter teardown).

Reply:
good catch I totally missed that. I'm going to add some tests to stress the finalizers.